clang/include/clang/Sema/BoundsAnalysis.h

```diff
@@ -253,6 +254,40 @@ namespace clang {
     // subscript or a pointer dereference.
     bool IsDeclOperand(const Expr *E);
 
+    // Make an expression uniform by moving all DeclRefExpr to the LHS and all
+    // IntegerLiterals to the RHS.
+    // @param[in] E is the expression which should be made uniform.
+    // @return A pair of expressions. The first contains all DeclRefExprs of E
+    // and the second contains all IntegerLiterals of E.
+    ExprPairTy SplitIntoBaseOffset(const Expr *E);
+
+    // An IntegerLiteral can either be int, +int or -int.
+    // @param[in] E is an expression.
+    // @return Whether E contains an IntegerLiteral.
+    bool IsIntegerOperand(const Expr *E) const;
+
+    // Get the APInt value from an expression representing an IntegerLiteral.
+    // @param[in] E represents an IntegerLiteral.
+    // @return Return the APInt value for E. Return APInt(0) if E is nullptr.
+    const llvm::APInt GetAPIntVal(const Expr *E) const;
```

Comment:
I think it would be better to use the function Expr::isIntegerConstantExpr instead of defining your own functions for recognizing constant expressions.  It does return an llvm::APSInt,  which I believe is what you want.  A clang IntegerLiteral expression uses an APInt (unsigned integer) to represent the bit pattern for the integer defined but the literal, but the bit pattern needs to be interpreted based on whether the *type* of the literal is a signed or unsigned integer type.   The function splitIntoBaseAndOffset in SemaBounds.cpp shows how to use isIntegerConstantExpr.